yorkie/src/main/kotlin/dev/yorkie/document/crdt/RHTPQMap.kt

```diff
@@ -0,0 +1,158 @@
+package dev.yorkie.document.crdt
+
+import dev.yorkie.document.time.TimeTicket
+import dev.yorkie.util.MaxPriorityQueue
+import dev.yorkie.util.PQNode
+import dev.yorkie.util.YorkieLogger
+
+/**
+ * RHTPQMap is replicated hash table with priority queue by creation time.
+ *
+ * @internal
+ */
+internal class RHTPQMap {
+    private val logTag = "RHTPQMap"
+
+    private val elementQueueMapByKey:
+        MutableMap<String, MaxPriorityQueue<PQNode<TimeTicket, CrdtElement>>> = mutableMapOf()
+
+    private val nodeMapByCreatedAt: MutableMap<TimeTicket, RHTPQMapNode> = mutableMapOf()
+
+    /**
+     * Sets the [value] using the given [key].
+     * If the object exists in [elementQueueMapByKey] by same [key] then return [CrdtElement], otherwise null.
+     */
+    fun set(key: String, value: CrdtElement): CrdtElement? {
+        var removed: CrdtElement? = null
+        val queue = elementQueueMapByKey[key]
+        if (queue != null && queue.size > 0) {
+            val node = queue.peek() as RHTPQMapNode
+            if (!node.isRemoved() && node.remove(value.createdAt)) {
+                removed = node.value
+            }
+        }
+        setInternal(key, value)
+        return removed
+    }
+
+    /**
+     * Sets the [value] using the given [key] internally.
+     */
+    private fun setInternal(key: String, value: CrdtElement) {
+        val queueMap = elementQueueMapByKey
+        if (!queueMap.contains(key)) {
+            queueMap[key] = MaxPriorityQueue()
+        }
+
+        val node = RHTPQMapNode.of(key, value)
+        val queue = queueMap[key]
+            ?: throw IllegalStateException("The MaxPriorityQueue by $key doesn't exist.")
+        queue.add(node)
+        nodeMapByCreatedAt[value.createdAt] = node
+    }
+
+    /**
+     * Deletes the Element in [nodeMapByCreatedAt] using the given key [createdAt],
+     * and TimeTicket will be removed by [executedAt].
+     * If the object exists in [nodeMapByCreatedAt] by same key [createdAt], it will return.
+     *
+     * @throws IllegalStateException if RHTPQMapNode doesn't exist.
+     */
+    fun delete(createdAt: TimeTicket, executedAt: TimeTicket): CrdtElement {
+        val nodeMap = nodeMapByCreatedAt
+        if (!nodeMap.contains(createdAt)) {
+            YorkieLogger.e(logTag, "fail to find $createdAt")
+        }
+
+        val node = nodeMap[createdAt]
+            ?: throw IllegalStateException("The RHTPQMapNode by $createdAt doesn't exist.")
+        node.remove(executedAt)
+        return node.value
+    }
+
+    /**
+     * Returns [RHTPQMapNode.strKey] of `node` based on creation time.
+     * The `node` will be found in [nodeMapByCreatedAt] using [createdAt]
+     *
+     * @throws IllegalStateException if RHTPQMapNode doesn't exist.
+     */
+    fun keyOf(createdAt: TimeTicket): String {
```

Comment:
@7hong13, @hyeongsik-won, and I decided to change `keyOf` to `subPathOf` to reveal its usage properly.

```suggestion
    fun subPathOf(createdAt: TimeTicket): String {
```
https://github.com/yorkie-team/yorkie-js-sdk/pull/391